### PR TITLE
Feature.extrusion

### PR DESCRIPTION
**Pull request by jbousquie**
Added two public extrusion methods : 
BASIC  
```javascript
BABYLON.Mesh.ExtrudeShape(name, shape, path, scale, rotation, scene)
```
* **shape** : the shape to be extruded, an array of successive Vector3,
* **path** : the path to extrude the shape along, an array of successive Vector3,
* **scale** : _default 1_, the value to scale the initial shape,
* **rotation** : _default 0_, the step value to rotate the shape at each path point.  

ADVANCED
```javascript
BABYLON.Mesh.ExtrudeShapeCustom(name, shape, path, scaleFunction, rotateFunction, ribbonCloseArray, ribbonClosePath, scene)
```
* **scaleFunction** : a custom javascript function. This function is called on each path point and is given the _i_ position of the point in the path and its _distance_ from the beginning of the path. It must return a _scale_ numeric value. This value will be the scaling applied to the shape drawn at the _i-th_  point.
* **rotationFunction** : a custom javascript function. This function is called on each path point and is given the _i_ position of the point in the path and its _distance_ from the beginning of the path. It must return a _rotation_ numeric value. This value will be the rotation applied to the shape drawn at the _i-th_  point.
* **ribbonCloseArray** : _default false_, boolean. The underlying ribbon _closeArray_  parameter. This can be used to close automatically a path with right normals computation.
* **ribbonClosePath** : _default false_, boolean. The underlying ribbon _closePath_  parameter. This can be used to close automatically a shape with right normals computation.

Of course the two last optional parameters _updatable?_ and _sideOrientation_ are avalaible but aren't described here, as well as _name_ and _scene_.

**NOTE :**
A shape is an array of successive Vector3. This means 2D or 3D shapes can be extruded as well.
The shape is to be designed in the local coordinate system knowing that the z-axis will be the extrusion path axis.
This means if you just want extrude 2D shapes, you will usually only define them in the xOy plane will all z coordinates valued to zero.
example : 
```javascript
var shape = [
  new BABYLON.Vector3(1, -1, 0),  
  new BABYLON.Vector3(1, 0, 0),
  new BABYLON.Vector3(0.6, 0.3, 0),
  new BABYLON.Vector3(0.8, 0.8, 0)
];
```
At last, shapes don't have to be centered in the local coordinate system.
A centered shape will extruded right on the path axis. An un-centered shape will be extruded offset from the path axis.

About the PR ...
Sorry for the bunch of js compiled files coming with ... just couldn't merge upstream/master unless they were commited :-(
Only **babylon.mesh.ts** is actually edited and modified.

### CODE REVIEW

**Comment by deltakosh** [comment]
Huge!! will need a doc :)

File: Babylon/Mesh/babylon.mesh.ts [hunk 1]
```diff
@@ -1167,6 +1167,52 @@
...
+                for (var p: number = 0; p < shape.length; p++) {
+                    var angleStep: number = rotate(i, distances[i]);
+                    var scaleRatio: number = scl(i, distances[i]);
+                    var rotationMatrix: Matrix = BABYLON.Matrix.RotationAxis(tangents[i], angle);
+                    var planed: Vector3 = ( (tangents[i].scale(shape[p].z)).add(normals[i].scale(shape[p].x)).add(binormals[i].scale(shape[p].y)) );
+                    var rotated: Vector3 = BABYLON.Vector3.TransformCoordinates(planed, rotationMatrix).scaleInPlace(scaleRatio).add(curve[i]);
+                    shapePath.push(rotated);
+                }
+                shapePaths.push(shapePath);
+                angle += angleStep;
```
Comment (deltakosh) [inline Babylon/Mesh/babylon.mesh.ts]: AngleStep is used outside the scope where it is defined. is this normal?

**Comment by jbousquie** [comment]
no, it's not... Bad port today from the original JS playground code due to a too short night :-(
Too bad the compiler didn't shout about this.

Just declare _angleStep_ in the upper level for-loop besides _shapePaths_, both evolves together/
I just notice _scaleRatio_  and _rotationMatrix_ can be set in the upper level for-loop also because they only depend upon var _i_ (path point position) and not upon var _p_ (shape points).
Pfffff  
Sorry for this... I prefer PR clean things.

Could you please fix it ?
Or I can re-commit and re-PR tomorrow if you prefer.

**Comment by jbousquie** [comment]
I'll write the doc with playground examples as soon as it is merged, no worries

**Comment by deltakosh** [comment]
I'll take care of this :)

**Comment by deltakosh** [comment]
It's online:)

**Comment by jbousquie** [comment]
thank you a lot :-D

I should have noticed by myself all those stuffs (wrong loop level, etc), pfff

**Comment by deltakosh** [comment]
My pleasure!!!

**Pull request closed without merging**